### PR TITLE
(WIP) Support of arguments in GTM & Segment sinks

### PR DESCRIPTION
**Pull request by dbMundada**
Adding details of data passed to GTM & Segment events in arguments.

### CODE REVIEW

File: src/main/scala/ai/privado/utility/Utilities.scala [hunk 1]
```diff
@@ -124,6 +148,209 @@ object Utilities {
...
+
+    processAssignmentNodes(assignmentNodes, "")
+    processSpreadOperatorNodes(spreadOperatorNodes, "")
+
+    if (node.name.equals("logSegmentEvent")) {
+      keyValueStructures += (("segmentData.page", "page"))
+      keyValueStructures += (("segmentArgs.meta", "analyticsData"))
+      keyValueStructures += (("segmentArgs.unataAppVersion ", "unataAppVersion"))
+      keyValueStructures += (("segmentArgs.unataMobileAppVersion", "unataMobileAppVersion"))
+      keyValueStructures += (("segmentArgs.unataPageViewId", "unataPageViewId"))
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/utility/Utilities.scala:347]: these values specific to an account, this shouldn't be part of privado-core

File: src/main/scala/ai/privado/utility/Utilities.scala [hunk 2]
```diff
@@ -124,6 +148,209 @@ object Utilities {
...
+      }
+
+      arrayCallNodes.foreach { bNode =>
+        handleBlockNode(bNode, Some(topLeftKey.getOrElse("") + ".[]"))
+      }
+
+      ""
+    }
+
+    val assignmentNodes     = node.argument.isBlock.astChildren.isCall.name("<operator>.assignment").l
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/utility/Utilities.scala:327]: Use `Operators` class

File: src/main/scala/ai/privado/utility/Utilities.scala [hunk 3]
```diff
@@ -124,6 +148,209 @@ object Utilities {
     storeForTagHelper(Constants.dbOperation, databaseDetails.dbOperation)
   }
 
+  def addArgumentsForGAPixelNode(node: Call, cpg: Cpg): List[(String, String)] = {
+    var keyValueStructures = ListBuffer.empty[(String, String)]
+
+    def processIdentifierNode(n: Identifier, topLeftKey: String): Unit = {
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/utility/Utilities.scala:154]: too many inner functions present inside `addArgumentsForGAPixelNode`

File: src/main/scala/ai/privado/utility/Utilities.scala [hunk 4]
```diff
@@ -124,6 +148,209 @@ object Utilities {
     storeForTagHelper(Constants.dbOperation, databaseDetails.dbOperation)
   }
 
+  def addArgumentsForGAPixelNode(node: Call, cpg: Cpg): List[(String, String)] = {
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/utility/Utilities.scala:151]: We can have this as part of `sink-> Utility`

File: src/main/scala/ai/privado/utility/Utilities.scala [hunk 5]
```diff
@@ -73,6 +86,17 @@ object Utilities {
 
   var ingressUrls = mutable.ListBuffer.empty[String]
 
+  def checkIfGTMOrSegment(ruleId: String): Boolean = {
+    val ruleList = List(
+      "ThirdParties.SDK.Segment.Analytics",
+      "ThirdParties.SDK.Pixel.Segment",
+      "ThirdParties.SDK.Google.TagManager",
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/utility/Utilities.scala:93]: These rules can be part of constants

File: src/test/scala/ai/privado/tagger/sink/SinkArgumentUtilityTest.scala [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+      ".payload.event"                          -> "\"productImpressions\"",
+      "products"                                -> "ANY",
+      "testPayload"                             -> "googlePayloads.logProductImpression.payload({})",
+      ".payload.pick.lodash"                    -> "lodash",
+      ".payload.\"ic_product_id\""              -> "product.\"ic_product_id\"",
+      ".payload.\"name\""                       -> "product.\"name\""
+    )
+    val encodedArgumentsList = cpg.tag.nameExact(Constants.arguments).value.headOption.getOrElse("{}")
+    val argumentsList        = SinkArgumentUtility.deserializedArgumentString(encodedArgumentsList)
+    cpg.tag.nameExact(Constants.id).value.headOption shouldBe Some(Constants.googleTagManagerPixelRuleId)
```
Comment (khemrajrathore) [inline src/test/scala/ai/privado/tagger/sink/SinkArgumentUtilityTest.scala:115]: It's not clear, which node is exactly getting tagged here

File: src/main/scala/ai/privado/exporter/ExporterUtility.scala [hunk 7]
```diff
@@ -201,7 +201,8 @@ object ExporterUtility {
     sizeOfList: Int = -1,
     messageInExcerpt: String = "",
     appCache: AppCache,
-    ruleCache: RuleCache
+    ruleCache: RuleCache,
+    arguments: String = ""
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/exporter/ExporterUtility.scala:205]: where is this parameter getting used?

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 8]
```diff
@@ -0,0 +1,291 @@
+package ai.privado.tagger.sink
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:1]: Any reason to move the above collection Taggers outside of the collection package?

Comment (dbMundada) [reply src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:1]: In taggers we have
- sinks
- collections
but initiallly direct `sink.collection` package created which not allowing me to create another folder package.

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+import io.shiftleft.codepropertygraph.generated.Cpg
+import io.shiftleft.codepropertygraph.generated.nodes.JavaProperty
+
+import scala.collection.mutable.ListBuffer
+import scala.util.matching.Regex
+import scala.collection.mutable.ListBuffer
+import java.io.{PrintWriter, ByteArrayInputStream, ByteArrayOutputStream, ObjectInputStream, ObjectOutputStream}
+import java.nio.file.Files
+
+object SinkArgumentUtility {
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:30]: A lot of things are done in this utility functions, please add comments explaining the use of each function and descriptive comment explaining the code functionality where ever required

File: src/main/scala/ai/privado/utility/Utilities.scala [hunk 10]
```diff
@@ -73,6 +75,17 @@ object Utilities {
...
+  def checkIfGTMOrSegment(ruleId: String): Boolean = {
+    val ruleList = List(
+      Constants.segmentPixelRuleId,
+      Constants.segmentAnalyticsRuleId,
+      Constants.googleTagManagerRuleId,
+      Constants.googleTagManagerPixelRuleId
+    )
+    val ruleExists = ruleList.contains(ruleId)
+    ruleExists
+  }
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/utility/Utilities.scala:87]: This can be re-written as an expression 
```
List(
      Constants.segmentPixelRuleId,
      Constants.segmentAnalyticsRuleId,
      Constants.googleTagManagerRuleId,
      Constants.googleTagManagerPixelRuleId
    ).contains(ruleId)

File: src/main/scala/ai/privado/utility/Utilities.scala [hunk 11]
```diff
@@ -151,6 +164,39 @@ object Utilities {
...
+        try {
+          val storeForTagHelper = storeForTag(builder, node, ruleCache) _
+          val callNode          = node.asInstanceOf[Call]
+          val argumentList: List[(String, String)] =
+            try {
+              SinkArgumentUtility.addArgumentsForGAPixelNode(callNode, cpg)
+            } catch {
+              case ex: Exception =>
+                logger.debug(s"An error occurred while processing arguments: ${ex.getMessage}")
+                List.empty[(String, String)] // Returning an empty list as a default value
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/utility/Utilities.scala:187]: better to also log some information like call Node code, arguments etc for debugging purpose

**Review by khemrajrathore — CHANGES_REQUESTED** [review]
- [ ] Multiple cases handled, but seeing only 1 test case, please ensure a test case is there for every case handled in code
- [ ] Add comments in the Utility function

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala (hunk 9 above)
Comment (dbMundada) [reply src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:30]: Added

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+    topLeftKey: String,
+    cpg: Cpg,
+    keyValueStructures: ListBuffer[(String, String)]
+  ): Unit = {
+    def updateLeftKey(code: String): String = {
+      if (topLeftKey.nonEmpty)
+        topLeftKey + "." + code
+      else
+        code
+    }
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:78]: [Optional] - This can be written as a expression
```
def updateLeftKey(code: String): String = if topLeftKey.nonEmpty then s"$topLeftKey.$code" else code

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+        (
+          getLeftKey(leftKey, topLeftKey, cpg, keyValueStructures),
+          Some(rightVal).collect {
+            case l: Literal         => l.code
+            case l: Identifier      => l.code
+            case l: FieldIdentifier => l.code
+            case l: Local           => l.code
+            case l: Call            => handleCallNode(l, topLeftKey, cpg, keyValueStructures).code
+            case l: Block =>
+              handleBlockNode(l, getLeftKey(leftKey, topLeftKey, cpg, keyValueStructures), cpg, keyValueStructures)
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:194]: There should a default case `case _ =>`

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+  private def getLeftKey(
+    leftKey: AstNode,
+    topLeftKey: Option[String],
+    cpg: Cpg,
+    keyValueStructures: ListBuffer[(String, String)]
+  ): Option[String] = {
+    val res = Some(leftKey)
+      .collect {
+        case c: Call    => handleCallNode(c, topLeftKey, cpg, keyValueStructures).astChildren.isFieldIdentifier.head
+        case l: Literal => l
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:163]: Add default

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+    topLeftKey: Option[String],
+    cpg: Cpg,
+    keyValueStructures: ListBuffer[(String, String)]
+  ): (Option[String], Option[String]) = {
+    objKeyVal match {
+      case leftKey :: rightVal :: _ =>
+        (
+          getLeftKey(leftKey, topLeftKey, cpg, keyValueStructures),
+          Some(rightVal).collect {
+            case l: Literal         => l.code
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:188]: `case x:@_ (Literal | Identifier)` something like this [Optional]

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+  ): Unit = {
+    spreadOperatorNodes.foreach { keyVal =>
+      val callNodes       = keyVal.astChildren.isCall.l
+      val childCallNodes  = callNodes.headOption.map(_.astChildren.l).getOrElse(List.empty)
+      val identifierNodes = keyVal.astChildren.isIdentifier.lastOption.l
+      callNodes.isCall.foreach { callN =>
+        handlePayloadCallNode(callN, topLeftKey, cpg, keyValueStructures)
+        handlePICKMethodCallNode(callN, topLeftKey, keyValueStructures)
+      }
+      (childCallNodes.collect { case n: Identifier => n } ++ identifierNodes.collect { case n: Identifier => n })
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:144]: `childCallNode.isIdentifier.l`

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+          .referencedMethod
+          .astChildren
+          .isBlock
+          .l
+        val returnBlockNode = internalBlockNode.astChildren.isReturn.astChildren.isBlock.l
+        handleBlockNode(blockNode.head, Some(topLeftKey), cpg, keyValueStructures)
+        handleBlockNode(returnBlockNode.head, Some(topLeftKey), cpg, keyValueStructures)
+      }
+    }
+  }
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:231]: There are lot of breaking changes in this code, like usage of head. Add exception handling

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+    processSpreadOperatorNodes(spreadOperatorNodes, topLeftKey.getOrElse(""), cpg, keyValueStructures)
+    childBlockNodes.foreach { bNode =>
+      handleBlockNode(bNode, topLeftKey, cpg, keyValueStructures)
+    }
+
+    arrayCallNodes.foreach { bNode =>
+      handleBlockNode(bNode, Some(topLeftKey.getOrElse("") + ".[]"), cpg, keyValueStructures)
+    }
+
+    ""
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:282]: Returning "" is not needed, this can be a unit method

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+    topLeftKey: String,
+    keyValueStructures: ListBuffer[(String, String)]
+  ): Unit = {
+    if (callNode.name.equals("pick")) {
+      var concatKey = ""
+      val keys      = callNode.astChildren.isBlock.astChildren.isCall.astChildren.isLiteral.code.l
+      val objName   = callNode.astChildren.isIdentifier.filter(i => !i.name.matches("this|pick")).name.l
+
+      if (objName.nonEmpty) {
+        concatKey = objName.head
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:299]: `val concatKey = if objName.nonEmpty then objName.head else ""`

File: src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+    objectOutputStream.writeObject(originalList.toMap)
+    objectOutputStream.close()
+    byteArrayOutputStream.toString("ISO-8859-1")
+  }
+
+  def deserializedArgumentString(serializedString: String): Map[String, String] = {
+    if (serializedString.isEmpty) {
+      Map.empty[String, String]
+    } else {
+      val byteArrayInputStream = new ByteArrayInputStream(serializedString.getBytes("ISO-8859-1"))
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:320]: Create a constant for `ISO-8859-1` and use

Comment (khemrajrathore) [inline src/main/scala/ai/privado/tagger/sink/SinkArgumentUtility.scala:320]: Check for robustness for this piece of code

**Review by khemrajrathore — CHANGES_REQUESTED** [review]
- [ ] Add exception handling and edge case handling for robustness
- [ ] Write modular unit test case, covering a specific style of implementation instead of a large 1 test case
- [ ] Refactor test case verbose to a self explanatory one

File: src/main/scala/ai/privado/model/Constants.scala [hunk 21]
```diff
@@ -145,6 +145,7 @@ object Constants {
   val EMPTY                         = "<empty>"
   val READ_WITH_BRACKETS            = " (Read)"
   val WRITE_WITH_BRACKETS           = " (Write)"
+  val BYE_ENCODING_CODE             = "ISO-8859-1"
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/model/Constants.scala:148]: rename to `BYTE_ENCODING_CODE`

File: src/test/scala/ai/privado/tagger/sink/SinkArgumentUtilityTest.scala [hunk 22]
```diff
@@ -17,136 +18,238 @@ import io.shiftleft.codepropertygraph.generated.Cpg
...
+          |    latitude: obj.latitude,
+          |    ...pick(product, ['id', 'name', 'ext_id', 'ic_product_id'])
+          |});
+          |""".stripMargin)
+      val expectedArgumentsMap = Map(
+        "latitude"           -> "obj.latitude",
+        ".\"id\""            -> "product.\"id\"",
+        ".\"ext_id\""        -> "product.\"ext_id\"",
+        "product"            -> "ANY",
+        ".pick.lodash"       -> "",
```
Comment (khemrajrathore) [inline src/test/scala/ai/privado/tagger/sink/SinkArgumentUtilityTest.scala:127]: Are these 2 values, `product` and `.pick.lodash` expected in the output?

File: src/test/scala/ai/privado/tagger/sink/SinkArgumentUtilityTest.scala [hunk 23]
```diff
@@ -17,136 +18,238 @@ import io.shiftleft.codepropertygraph.generated.Cpg
...
+
+      val gaSinkNode = cpg.call.name("push").head
+      gaSinkNode.tag.nameExact(Constants.id).value.headOption shouldBe Some(Constants.googleTagManagerPixelRuleId)
+      val encodedArgumentsList = gaSinkNode.tag.nameExact(Constants.arguments).value.headOption.getOrElse("{}")
+      val argumentsList        = SinkArgumentUtility.deserializedArgumentString(encodedArgumentsList)
+      // Iterate over each key-value pair in the argumentsMap
+      expectedArgumentsMap.foreach { case (key, value) =>
+        // Retrieve the corresponding value from the argumentsList
+        val argumentsListValue = argumentsList.getOrElse(key, "")
+        assert(argumentsListValue == value, s"Expected value '$value' for key '$key', but got '$argumentsListValue'")
```
Comment (khemrajrathore) [inline src/test/scala/ai/privado/tagger/sink/SinkArgumentUtilityTest.scala:140]: I see these 4-5 lines being repeated across, create a method and use that instead

File: src/test/scala/ai/privado/tagger/sink/SinkArgumentUtilityTest.scala [hunk 24]
```diff
@@ -17,136 +18,238 @@ import io.shiftleft.codepropertygraph.generated.Cpg
...
+
+      // Iterate over each key-value pair in the argumentsMap
+      expectedArgumentsMap.foreach { case (key, value) =>
+        // Retrieve the corresponding value from the argumentsList
+        val argumentsListValue = argumentsList.getOrElse(key, "")
+        assert(argumentsListValue == value, s"Expected value '$value' for key '$key', but got '$argumentsListValue'")
+      }
+    }
+
+//    "Check spread operator call node arguments" in {
```
Comment (khemrajrathore) [inline src/test/scala/ai/privado/tagger/sink/SinkArgumentUtilityTest.scala:195]: To ignore a test case we can write `ignore` instead of `in`

File: src/test/scala/ai/privado/tagger/sink/SinkArgumentUtilityTest.scala (hunk 22 above)
Comment (dbMundada) [reply src/test/scala/ai/privado/tagger/sink/SinkArgumentUtilityTest.scala:115]: For now yes, because of nested structure we created they are coming.
Will work over it as further customer requirements comes in.

**Review by khemrajrathore — APPROVED** [review]
LGTM

**Comment by pandurangpatil** [comment]
attaching code coverage report for SinkgArgumentUtility which contains the majority of the changes.

image
